File: Greedy/Dijkstra's Shortest Path/code.cpp
#include "algorithm-visualizer.h"
#include <iostream>
#include <vector>

GraphTracer				tracer("GraphTracer");
Array1DTracer			tracerS("Array1DTracer");
LogTracer				logger("LogTracer");
Randomize::Graph<int>	graph(5, 1);
int G[5][5];
int _G[25];
int MAX_VALUE = 2147483647;
int S[5];

void InitDijkstraShortestPath()
{
	Layout::setRoot(VerticalLayout({ tracer,tracerS ,logger }));
	tracer.directed(false);
	tracer.weighted();
	tracer.log(logger);
	graph.directed(false);
	graph.weighted();
	graph.fill(_G);
	for (int i = 0; i < 5; i++) {
		for (int j = 0; j < 5; j++) {
			G[i][j] = _G[i * 5 + j];
		}
	}
	tracer.set(G);
	for (int i = 0; i < 5; i++)
		S[i] = MAX_VALUE;
	tracerS.set(S);
	Tracer::delay();
}

void Dijkstra(int start, int end) {
	int minIndex;
	int minDistance;
	std::vector<int> D;
	for (int i = 0; i < 5; i++) D.push_back(false);
	S[start] = 0; 
	tracerS.patch(start, S[start]);
	Tracer::delay();
	tracerS.depatch(start);
	tracerS.select(start);
	int k = 5;
	while (k--) {
		minDistance = MAX_VALUE;
		for (int i = 0; i < 5; i++) {
			if (S[i] < minDistance && !D[i]) {
				minDistance = S[i];
				minIndex = i;
			}
		}
		if (minDistance == MAX_VALUE) break; 
		D[minIndex] = true;
		tracerS.select(minIndex);
		tracer.visit(minIndex);
		Tracer::delay();
		for (int i = 0; i < 5; i++) {
			if (G[minIndex][i] && S[i] > S[minIndex] + G[minIndex][i]) {
				S[i] = S[minIndex] + G[minIndex][i];
				tracerS.patch(i, S[i]);
				tracer.visit(i, minIndex, S[i]);
				Tracer::delay();
				tracerS.depatch(i);
				tracer.leave(i, minIndex);
				Tracer::delay();
			}
		}
		tracer.leave(minIndex);
		Tracer::delay();
	}
	if (S[end] == MAX_VALUE) {
		logger.println("there is no path from " + std::to_string(start) + " to " 
			+std::to_string(end));
	}
	else {
		logger.println("the shortest path from " + std::to_string(start) + " to "+
			std::to_string(end) + " is " + std::to_string(S[end]));
	}
}

int main()
{
	InitDijkstraShortestPath();
	Randomize::Integer randInteger(0, 4);
	int s = randInteger.create();
	int e;
	do {
		e = randInteger.create();
	} while (s == e);
	logger.println("finding the shortest path from "+ std::to_string(s) + " to " 
		+std::to_string(e));
	Tracer::delay();
	Dijkstra(s, e);
}

File: Greedy/Dijkstra's Shortest Path/code.java
import org.algorithm_visualizer.*;

public class Main {
    //visualizer {
	GraphTracer tracer = new GraphTracer().directed(false).weighted();
	Array1DTracer tracerS = new Array1DTracer();
	LogTracer logger = new LogTracer();
	Object[][] G = new Randomize.Graph(5, 1).directed(false).weighted().create();
	//}
	public static Integer MAX_VALUE = Integer.MAX_VALUE;
	public static Integer[] S = new Integer[5]; // S[end] returns the distance from start node to end node
	
	Main() {
	    //visualizer {
		Layout.setRoot(new VerticalLayout(new Commander[] {tracer, tracerS, logger}));
		tracer.log(logger);
		tracer.set(G);
		//}
		for (int i = 0; i < G.length; i++) 
		    S[i] = MAX_VALUE;
		//visualizer {
		tracerS.set(S);
		Tracer.delay();
		//}
		//start node : s = 0~ G.length 사이 랜덤값 {
		Integer s = new Randomize.Integer(0, G.length - 1).create();
		//}
		Integer e; // e = end node
		do {
		    //end node : e = 0~ G.length 사이 랜덤값 {
		  e = new Randomize.Integer(0, G.length - 1).create();
		  //}
		} while (s == e);
		//visualizer {
		logger.println("finding the shortest path from " + Integer.toString(s) + " to " + Integer.toString(e));
		Tracer.delay();
		//}
		Dijkstra(s, e);
	}
	
	void Dijkstra(Integer start, Integer end) {
	  int minIndex = 0;
	  int minDistance;
	  boolean[] D = new boolean[5]; // D[i] indicates whether the i-th node is discovered or not
	  for (int i = 0; i < G.length; i++) D[i] = false;
	  S[start] = (int)0; // Starting node is at distance 0 from itself
	  //visualizer {
	  tracerS.depatch(start, S[start]);
	  Tracer.delay();
	  tracerS.depatch(start);
	  tracerS.select(start);
	  //}
	  int k = G.length;
	  boolean temp = (k-- != 0);
	  while (temp) {
	    // Finding a node with the shortest distance from S[minIndex]
	    minDistance = MAX_VALUE;
	    for (int i = 0; i < G.length; i++) {
	      if (S[i] < minDistance && !D[i]) {
	        minDistance = S[i];
	        minIndex = i;
	      }
	    }
	    if (minDistance == MAX_VALUE) break; // If there is no edge from current node, jump out of loop
	    D[minIndex] = true;
	    //visualizer {
	    tracerS.select(minIndex);
	    tracer.visit(minIndex);
	    Tracer.delay();
	    //}
	    // For every unvisited neighbour of current node, we check
	    // whether the path to it is shorter if going over the current node
	    for (int i = 0; i < G.length; i++) {
	    	if (G[minIndex][i] != null && S[i] > S[minIndex] + (Integer)G[minIndex][i])
	    	{
	    		S[i] = S[minIndex] + (Integer)G[minIndex][i];
	    		//visualizer {
		        tracerS.patch(i, S[i]);
		        tracer.visit(i, minIndex, S[i]);
		        Tracer.delay();
		        tracerS.depatch(i);
		        tracer.leave(i, minIndex);
		        Tracer.delay();
		        //}
	    	}
	    }
	    //visualizer {
	    tracer.leave(minIndex);
	    Tracer.delay();
	    //}
	  }
	  if (S[end] == MAX_VALUE) {
	    logger.println("there is no path from " + Integer.toString(start) +  " to " +  Integer.toString(end));
	  } else {
	    logger.println("the shortest path from " + Integer.toString(start) + " to " + Integer.toString(end) + " is " +  Integer.toString(S[end]));
	  }
	}

    public static void main(String[] args) {
        new Main();
    }
}
